### ISSUE TITLE
Save Best Model option for eval.py

### ISSUE THREAD

**Issue by mbenami** [tensorflow/models#4941]
- **What is the top-level directory of the model you are using**: object_detection
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: N/a
- **TensorFlow installed from (source or binary)**:source 
- **TensorFlow version (use command below)**:N/a
- **Bazel version (if compiling from source)**:N/a
- **CUDA/cuDNN version**:N/a
- **GPU model and memory**:N/a
- **Exact command to reproduce**:N/a

New feature, 
Hi I know that in keras, when you train there is an option `save_best` for saving the best model evaluation depend on the matric you want (`loss`, `acc`, or custom )

is there an option to add this when using eval.py for evaluating the train model? 
the matric can be one of the `metrics_set` in the eval_config 
that can save space when training for long time and you don't want to save all the train checkpoint 

Thanks.

 

Please go to Stack Overflow for help and support:

http://stackoverflow.com/questions/tagged/tensorflow

Also, please understand that many of the models included in this repository are experimental and research-style code. If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **What is the top-level directory of the model you are using**:
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **TensorFlow installed from (source or binary)**:
- **TensorFlow version (use command below)**:
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by qlzh727** [comment]
I cannot speak for object detection team, but I think currently the model is not implemented by keras. Probably this can be a nice feature if object detection team migrate their implementation from estimator to keras.

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.

**Comment by mbenami** [comment]
I found this issue https://github.com/tensorflow/tensorflow/issues/8658
that says how to copy the best checkpoint 

but I could not find any way to export a frozen_inference_graph.pb file when I get a better evaluation 
the only way I succeed to do this is to call python export_inference_graph.py as cmd 
```
os.(" python object_detection/export_inference_graph.py --input_type=image_tensor 
....
")
```
with setting 
```
os.environ["CUDA_VISIBLE_DEVICES"]="-1"
```
in the script, so I would not get OOM 
but I could not use this workaround when I train on the cloud with 
```
gcloud ml-engine jobs submit training ...
```

to summary 
it would be great if someone knows how I can export frozen_graph (with input_type as image_tensor) 
or change 
Thanks

**Comment by mbenami** [comment]
Update :
with the help from the issue, I mentioned earlier here 

I was able to export save_model.pb with `input_type=image_tensor`
so for detection my API stays the same . 
you can use the `BestExporter` form `tf.estimator.BestExporter`
with the folwing `serving_input_receiver_fn` 

```
def serving_input_receiver_fn(model_config):

  def _transform_input_data(tensor_dict,
                           model_preprocess_fn,
                           image_resizer_fn,
                           retain_original_image=True):
    out_tensor_dict = tensor_dict.copy()
    if retain_original_image:
        out_tensor_dict[fields.InputDataFields.original_image] = tf.cast(
          image_resizer_fn(
            out_tensor_dict[fields.InputDataFields.image],None)[0],
            tf.uint8)
    # Apply model preprocessing ops and resize instance masks.
    image = out_tensor_dict[fields.InputDataFields.image]
    preprocessed_resized_image, true_image_shape = model_preprocess_fn(tf.expand_dims(tf.cast(image, dtype=tf.float32), axis=0))
    preprocessed_shape = tf.shape(preprocessed_resized_image)
    new_height, new_width = preprocessed_shape[1], preprocessed_shape[2]
    out_tensor_dict[fields.InputDataFields.image] = tf.squeeze(preprocessed_resized_image, axis=0)
    out_tensor_dict[fields.InputDataFields.true_image_shape] = tf.squeeze(true_image_shape, axis=0)
    return out_tensor_dict
    
  def _serving_input_receiver_fn(params=None):
    del params
    example = tf.placeholder(dtype=tf.uint8, shape=(None, None, 3) ,name='image_tensor')
    num_classes = config_util.get_number_of_classes(model_config)
    model_preprocess_fn = inputs.INPUT_BUILDER_UTIL_MAP['model_build'](
        model_config, is_training=False).preprocess

    image_resizer_config = config_util.get_image_resizer_config(model_config)
    image_resizer_fn = image_resizer_builder.build(image_resizer_config)
    transform_fn = functools.partial(_transform_input_data,
                                      model_preprocess_fn=model_preprocess_fn,
                                      image_resizer_fn=image_resizer_fn)
    input_dict = transform_fn({fields.InputDataFields.image:example})
    images = tf.cast(input_dict[fields.InputDataFields.image], dtype=tf.float32)
    images = tf.expand_dims(images, axis=0)
    true_image_shape = tf.expand_dims(
        input_dict[fields.InputDataFields.true_image_shape], axis=0)

    return tf.estimator.export.ServingInputReceiver(
        features={
            fields.InputDataFields.image: images,
            fields.InputDataFields.true_image_shape: true_image_shape},
        receiver_tensors={'image_tensor': example})

  return _serving_input_receiver_fn

```
